File: leetcode/54. Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int mx = matrix.size(),
            my = matrix[0].size();
        vector<int> ans;
        int sx = 0, sy = 0;
        while (sx < mx and sy < my) {
            // right
            for (int i = sy; i < my; i++) {
                ans.push_back(matrix[sx][i]);
            }
            sx++;
            // down
            for (int i = sx; i < mx; i++) {
                ans.push_back(matrix[i][my - 1]);
            }
            my--;
            // left
            // check if there is a row left
            if (sx < mx) {
                for (int i = my - 1; i >= sy; i--) {
                    ans.push_back(matrix[mx - 1][i]);
                }
                mx--;
            }
            // up
            // check if there is a column left
            if (sy < my) {
                for (int i = mx - 1; i >= sx; i--) {
                    ans.push_back(matrix[i][sy]);
                }
                sy++;
            }
        }
        return ans;
    }
};

File: leetcode/54. Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        x = len(matrix)
        y = len(matrix[0])
        ans = []
        i = 0
        j = 0
        while i < x and j < y:
            # right
            for k in range(j, y):
                ans.append(matrix[i][k])
            i += 1
            # down
            for k in range(i, x):
                ans.append(matrix[k][y-1])
            y -= 1
            # left
            if i < x:
                for k in range(y-1, j-1, -1):
                    ans.append(matrix[x-1][k])
                x -= 1
            # up
            if j < y:
                for k in range(x-1, i-1, -1):
                    ans.append(matrix[k][j])
                j += 1
        return ans
    
